extensions/src/main/java/org/pentaho/platform/web/http/api/resources/utils/SystemUtils.java

```diff
@@ -21,16 +21,89 @@
+import org.pentaho.platform.api.engine.IUserRoleListService;
+import org.pentaho.platform.engine.core.system.PentahoSessionHolder;
 import org.pentaho.platform.engine.core.system.PentahoSystem;
+import org.pentaho.platform.repository2.unified.ServerRepositoryPaths;
+import org.pentaho.platform.repository2.unified.jcr.JcrTenantUtils;
 import org.pentaho.platform.security.policy.rolebased.actions.AdministerSecurityAction;
+import org.pentaho.platform.security.policy.rolebased.actions.PublishAction;
 import org.pentaho.platform.security.policy.rolebased.actions.RepositoryCreateAction;
 import org.pentaho.platform.security.policy.rolebased.actions.RepositoryReadAction;
 
+import java.util.Collections;
+import java.util.List;
+
 public class SystemUtils {
+
+  private static final Log logger = LogFactory.getLog( SystemUtils.class );
+
   public static boolean canAdminister() {
     IAuthorizationPolicy policy = PentahoSystem.get( IAuthorizationPolicy.class );
     return policy.isAllowed( RepositoryReadAction.NAME ) && policy.isAllowed( RepositoryCreateAction.NAME )
         && ( policy.isAllowed( AdministerSecurityAction.NAME ) );
   }
+
+  public static boolean canUpload( String uploadDir ) {
+    IAuthorizationPolicy policy = PentahoSystem.get( IAuthorizationPolicy.class );
+
+    //check if we are admin or have publish permission
+    boolean isAdmin = policy.isAllowed( RepositoryReadAction.NAME ) && policy.isAllowed( RepositoryCreateAction.NAME )
+      && ( policy.isAllowed( AdministerSecurityAction.NAME ) || policy.isAllowed( PublishAction.NAME ) );
+
+    //the user does not have admin or publish permission, so we will check if the user imports to their home folder
+    if ( !isAdmin && !StringUtils.isEmpty( uploadDir ) ) {
+      return validateAccessToHomeFolder( uploadDir );
+    }
+
+    return isAdmin;
+  }
+
+  public static boolean canDownload( String downloadDir ) {
+    IAuthorizationPolicy policy = PentahoSystem.get( IAuthorizationPolicy.class );
+
+    IUserRoleListService userRoleListService = PentahoSystem.get( IUserRoleListService.class );
+    String tenantedUserName = PentahoSessionHolder.getSession().getName();
+    List<String> tenantedUserRoles = userRoleListService.getRolesForUser( JcrTenantUtils.getUserNameUtils().getTenant( tenantedUserName ), tenantedUserName );
+
+    //check if we are admin or have download-roles
+    boolean isAdminOrHaveDownloadActionRole = policy.isAllowed( RepositoryReadAction.NAME )
+      && policy.isAllowed( RepositoryCreateAction.NAME )
+      && ( policy.isAllowed( AdministerSecurityAction.NAME )
+      || !Collections.disjoint( tenantedUserRoles, PentahoSystem.getDownloadRolesList() ) );
+
+    //the user does not have admin or download-roles assigned, so we will check if the user downloads from their home folder
+    if ( !isAdminOrHaveDownloadActionRole && !StringUtils.isEmpty( downloadDir ) ) {
+      return validateAccessToHomeFolder( downloadDir );
+    }
```

Comment:
Might be more efficient to check this first so we skip all the role and permission checks if downloadDir is under user home folder.